jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/DulmageMendelsohnDecomposition.java

```diff
@@ -0,0 +1,302 @@
+/*
+ * (C) Copyright 2018-2018, by CAE Tech Limited and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.connectivity;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.jgrapht.Graph;
+import org.jgrapht.alg.interfaces.MatchingAlgorithm.Matching;
+import org.jgrapht.alg.interfaces.StrongConnectivityAlgorithm;
+import org.jgrapht.alg.matching.HopcroftKarpMaximumCardinalityBipartiteMatching;
+import org.jgrapht.graph.AsSubgraph;
+import org.jgrapht.graph.DefaultDirectedGraph;
+import org.jgrapht.graph.DefaultEdge;
+import org.jgrapht.graph.EdgeReversedGraph;
+import org.jgrapht.graph.builder.GraphBuilder;
+import org.jgrapht.traverse.DepthFirstIterator;
+
+/**
+ * The Dulmage-Mendelsohn Decomposition
+ * (https://en.wikipedia.org/wiki/Dulmage%E2%80%93Mendelsohn_decomposition)
+ * partitions a bipartite graph into subsets. One subset is dominated by the
+ * first partition, one by the second partition, and the third is the remaining
+ * vertices.
+ *
+ * A fine decomposition can be performed that additionally performs a
+ * strongly-connected-components algorithm on the remaining subset.
+ * 
```

Comment:
Per https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation could you add the complexity info?

Reply:
@jsichi Thanks for the review, I have removed the BipartiteInspector (I hadn't noticed the method in GraphTests!) and added a complexity